### PR TITLE
MEN-5223 Monitor: streamline logs and fixes

### PR DESCRIPTION
**Pull request by merlin-northern**
ChangeLog:title
Signed-off-by: Peter Grzybowski <peter@northern.tech>

### CODE REVIEW

File: 01.Get-started/05.Monitor/docs.md [hunk 1]
```diff
@@ -231,6 +231,34 @@ readlink /etc/mender-monitor/monitor.d/enabled/service_mender-connect.sh
...
+```
+> ```bash
+> # This file was autogenerated by Monitoring Utilities based on the configuration
+> SERVICE_NAME="my_service_logs"
+> LOG_PATTERN="Exited with code \d+"
+> LOG_FILE="@journalctl -u my-service -f"
+> LOG_PATTERN_EXPIRATION=255
+> ```
+
+Which will trigger a critical alert everytime _Exited with code n_ (where n is any decimal number) shows up
```
Comment (lluiscampos) [inline 01.Get-started/05.Monitor/docs.md:256]:
```suggestion
Which will trigger a critical alert every time _Exited with code n_ (where n is any decimal number) shows up
```

File: 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md [hunk 2]
```diff
@@ -83,6 +83,50 @@ the [Perl-compatible regular expressions](https://www.pcre.org/).
...
+cases you can provide wrapper scripts around some of more complicated commands.
+
+### Log pattern expiration
+
+Once the pattern shows up in the logs, Mender Monitor add-on will send a critical alert.
+Technically there would be no way out of this situation; the device would stay
+in the "critical" state forever. Monitor add-on provides a way to send
+the _"OK"_ alert after a period of time, either provided with `LOG_PATTERN_EXPIRATION`
+or by the last argument to the `mender-monitorctl create log name pattern file expiration`
+command.
```
Comment (lluiscampos) [inline 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md:127]: Missing new line here

Comment (lluiscampos) [inline 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md:127]: It says "either" this or that, but both cases are the same.
```suggestion
the _"OK"_ alert after a period of time, defined by `LOG_PATTERN_EXPIRATION` in the check. This variable is set via
the last argument to the `mender-monitorctl create log name pattern file expiration`
command.
```

Something like that...

File: 01.Get-started/05.Monitor/docs.md [hunk 3]
```diff
@@ -231,6 +231,34 @@ readlink /etc/mender-monitor/monitor.d/enabled/service_mender-connect.sh
...
+in the logs for `my-service` systemd service. In the above example there is going to be an automatical cancelation of the critical
+condition (sending of an "OK" alert) after 255 consequtive seconds during which the pattern does not appear in the logs.
+This is what the last and optional argument to `mender-monitorctl` stands for.
+
```
Comment (lluiscampos) [inline 01.Get-started/05.Monitor/docs.md:260]: At the end here I would add a reference for more information to go to `09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md`

**Review by lluiscampos — CHANGES_REQUESTED** [review]

File: 01.Get-started/05.Monitor/docs.md (hunk 3 above)
Comment (merlin-northern) [reply 01.Get-started/05.Monitor/docs.md:260]: added:
```md
For more information please refer
to the [Monitoring subsystems](../../09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md) section.
```

File: 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md (hunk 2 above)
Comment (merlin-northern) [reply 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md]: makes sense. fixed

Comment (merlin-northern) [reply 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md:132]: fixed

File: 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md [hunk 4]
```diff
@@ -83,6 +83,50 @@ the [Perl-compatible regular expressions](https://www.pcre.org/).
...
+One example of a command that can output data to standard error is `docker logs`.
+In case you want to get alerts based on the patterns that maybe present
+in the standard error stream, you can configure the check as follows:
+
+```bash
+mender-monitorctl create log docker_logs "Exception \d+-\w+ found" "@docker logs my_service 2>&1"
+```
+
+With the above you can configure alerting based on any source. In extreme
+cases you can provide wrapper scripts around some of more complicated commands.
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md]: fixed. I dropped "some" from the "You can also provide wrapper scripts for some more advance commands" and changed advance -> advanced

File: 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md [hunk 5]
```diff
@@ -83,6 +83,50 @@ the [Perl-compatible regular expressions](https://www.pcre.org/).
...
+```
+
+Where `/opt/bin/logs-secure.sh` is any executable command printing to standard output.
+
+As we saw in the [Get started](../../../01.Get-started/05.Monitor/docs.md) section
+the `journalctl` command can be a source of logs, but we do not limit the mechanism
+to any particular command. The only requirement is: the log subsystem will parse
+the standard output.
+
+#### docker logs
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md]: fixed. it was meant more as command `docker logs` but capital letter is better here.

File: 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md [hunk 6]
```diff
@@ -83,6 +83,50 @@ the [Perl-compatible regular expressions](https://www.pcre.org/).
...
+```bash
+mender-monitorctl create log custom_logs "User logged out due to security policy: \w+" "@/opt/bin/logs-secure.sh"
+```
+
+Where `/opt/bin/logs-secure.sh` is any executable command printing to standard output.
+
+As we saw in the [Get started](../../../01.Get-started/05.Monitor/docs.md) section
+the `journalctl` command can be a source of logs, but we do not limit the mechanism
+to any particular command. The only requirement is: the log subsystem will parse
+the standard output.
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md]: fixed.

File: 01.Get-started/05.Monitor/docs.md [hunk 7]
```diff
@@ -231,6 +231,34 @@ readlink /etc/mender-monitor/monitor.d/enabled/service_mender-connect.sh
...
+> ```bash
+> # This file was autogenerated by Monitoring Utilities based on the configuration
+> SERVICE_NAME="my_service_logs"
+> LOG_PATTERN="Exited with code \d+"
+> LOG_FILE="@journalctl -u my-service -f"
+> LOG_PATTERN_EXPIRATION=255
+> ```
+
+Which will trigger a critical alert everytime _Exited with code n_ (where n is any decimal number) shows up
+in the logs for `my-service` systemd service. In the above example there is going to be an automatical cancelation of the critical
```
[the remark this answers is not in the document]
Comment (merlin-northern) [reply 01.Get-started/05.Monitor/docs.md]: fixed.

File: 01.Get-started/05.Monitor/docs.md (hunk 1 above)
Comment (merlin-northern) [reply 01.Get-started/05.Monitor/docs.md]: fixed.

File: 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md [hunk 8]
```diff
@@ -83,6 +83,56 @@ the [Perl-compatible regular expressions](https://www.pcre.org/).
...
+Technically there would be no way out of this situation; the device would stay
+in the "critical" state forever. Monitor add-on provides a way to send
+the _"OK"_ alert after a period of time defined by `LOG_PATTERN_EXPIRATION`
+in the check. This variable is set via the last argument
+to the `mender-monitorctl create log name pattern file expiration` command.
+
+
+either provided with `LOG_PATTERN_EXPIRATION`
+or by the last argument to the `mender-monitorctl create log name pattern file expiration`
+command.
```
Comment (lluiscampos) [inline 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md:132]: Something estrange here now. Maybe:
```suggestion
in the "critical" state forever. Monitor add-on provides a way to send
the _"OK"_ alert after a period of time defined by `LOG_PATTERN_EXPIRATION`
in the check. This variable is defined either directly on the check
or by the last argument to the `mender-monitorctl create log name pattern file expiration`
command.
```

**Review by lluiscampos — APPROVED** [review]
Looks good. Only one sentence that does not fully make sense now.

File: 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md (hunk 8 above)
Comment (merlin-northern) [reply 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md:132]: ho crap! sorry! old misplaced line, forgot to delete. shame on me.

Comment (lluiscampos) [reply 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md]: No problem. That is the purpose of reviews ;)

Comment (merlin-northern) [reply 09.Add-ons/20.Monitor/20.Monitoring-subsystems/docs.md]: thank you Sir!

**Review by lluiscampos — APPROVED** [review]

**Pull request merged**